### PR TITLE
[IoTDB-2991] Equal size bucket sampling UDFs: EQUAL_SIZE_BUCKET_OUTLIER_SAMPLE

### PR DESCRIPTION
**Pull request by ZhanGHanG9991**
## Description
Equal size bucket sampling UDFs: EQUAL_SIZE_BUCKET_OUTLIER_SAMPLE

<hr>

This PR has:
- [x] been self-reviewed.
    - [ ] concurrent read
    - [ ] concurrent write
    - [ ] concurrent read and write 
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. 
- [x] added or updated version, __license__, or notice information
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious 
  for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold 
  for code coverage.
- [x] added integration tests.
- [ ] been tested in a test IoTDB cluster.

<hr>

### CODE REVIEW

File: integration/src/test/java/org/apache/iotdb/db/integration/IoTDBUDTFBuiltinFunctionIT.java [hunk 1]
```diff
@@ -767,4 +767,102 @@ public void testEqualBucketSampleForM4() {
...
+                  proportionValue,
+                  "stendis",
+                  2));
+      int columnCount2 = resultSet2.getMetaData().getColumnCount();
+      assertEquals(1 + 2, columnCount2);
+      for (int i = 0; i < 10; i++) {
+        resultSet2.next();
+        System.out.println(resultSet2.getInt(2) + " " + resultSet2.getInt(3));
+        //        assertEquals(ANSWER1[i], resultSet2.getInt(2));
+        //        assertEquals(ANSWER2[i], resultSet2.getInt(3));
```
Comment (ZhanGHanG9991) [inline integration/src/test/java/org/apache/iotdb/db/integration/IoTDBUDTFBuiltinFunctionIT.java:862]: 在这里出现了一个很奇怪的问题。在输出“pass”上面的两个单独的select语句都有符合预期的输出，但是两个语句合在一起结果却不同。不知道是为什么？

File: integration/src/test/java/org/apache/iotdb/db/integration/IoTDBUDTFBuiltinFunctionIT.java [hunk 2]
```diff
@@ -767,4 +767,102 @@ public void testEqualBucketSampleForM4() {
...
+      }
```
Comment (ZhanGHanG9991) [inline integration/src/test/java/org/apache/iotdb/db/integration/IoTDBUDTFBuiltinFunctionIT.java:863]: 在这里出现了一个很奇怪的问题。在输出System.out.println("-------------pass--------------")前，两个select 语句单独执行有符合预期的结果，但是如果将两个select并在一起执行，输出不符合预期，不清楚是为什么？

Comment (ZhanGHanG9991) [reply integration/src/test/java/org/apache/iotdb/db/integration/IoTDBUDTFBuiltinFunctionIT.java]: 问题已解决，时间序列没有对齐。

File: docs/UserGuide/Query-Data/Select-Expression.md [hunk 3]
```diff
@@ -668,6 +668,80 @@ Result:
...
+
+Parameter Description
+- `proportion`: sampling ratio
+- `number`: the number of samples in each bucket, default `3`
+- `type`: outlier sampling method, the value is
+  - `avg`: Take the average of the data points in the bucket, and find the `top number` farthest from the average according to the sampling ratio
+  - `stendis`: Take the vertical distance between each data point in the bucket and the first and last data points of the bucket to form a straight line, and according to the sampling ratio, find the `top number` with the largest distance
+  - `cos`: Set a data point in the bucket as b, the data point on the left of b as a, and the data point on the right of b as c, then take the cosine value of the angle between the ab and bc vectors. The larger the angle, the more likely it is an outlier. Find the `top number` with the smallest cos value
+  - `prenextdis`: Set a data point in the bucket as b, the data point to the left of b as a, and the data point to the right of b as c, then take the cosine value of the angle between the ab and bc vectors. The larger the angle, the more likely it is an outlier. Find the `top number` with the smallest cos value
```
Comment (SteveYurongSu) [inline docs/UserGuide/Query-Data/Select-Expression.md:685]: Is it the same with `cos`?

File: docs/zh/UserGuide/Query-Data/Select-Expression.md [hunk 4]
```diff
@@ -672,6 +672,82 @@ Total line number = 12
...
+| EQUAL_SIZE_BUCKET_OUTLIER_SAMPLE   | INT32 / INT64 / FLOAT / DOUBLE | `proportion`取值范围为`(0, 1]`，默认为`0.1`</br>`type`取值为`avg`或`stendis`或`cos`或`prenextdis`，默认为`avg`</br>`number`取值应大于0，默认`3`| INT32 / INT64 / FLOAT / DOUBLE | 返回符合采样比例和桶内采样个数的等分桶离群值采样                |
+
+参数说明
+- `proportion`: 采样比例
+  - `number`: 每个桶内的采样个数，默认`3`
+- `type`: 离群值采样方法，取值为
+    - `avg`: 取桶内数据点的平均值，并根据采样比例，找到距离均值最远的top number个
+    - `stendis`: 取桶内每一个数据点距离桶的首末数据点连成直线的垂直距离，并根据采样比例，找到距离最大的top number个
+    - `cos`: 设桶内一个数据点为b，b左边的数据点为a，b右边的数据点为c，则取ab与bc向量的夹角的余弦值，值越小，说明形成的角度越大，越可能是异常值。找到cos值最小的top number个
+    - `prenextdis`: 设桶内一个数据点为b，b左边的数据点为a，b右边的数据点为c，则取ab与bc向量的夹角的余弦值，值越小，说明形成的角度越大，越可能是异常值。找到cos值最小的top number个
```
Comment (SteveYurongSu) [inline docs/zh/UserGuide/Query-Data/Select-Expression.md:689]: Is it the same with `cos`?

File: server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketOutlierSample.java [hunk 5]
```diff
@@ -0,0 +1,1067 @@
...
+
+      double avg = 0;
+      PriorityQueue<Pair<Integer, Double>> pq =
+          new PriorityQueue<>(number, Comparator.comparing(o -> o.right));
+      for (int i = 0; i < windowSize; i++) {
+        avg += rowWindow.getRow(i).getInt(0);
+      }
+      avg /= windowSize;
+      for (int i = 0; i < windowSize; i++) {
+        double value = rowWindow.getRow(i).getInt(0) - avg;
```
Comment (SteveYurongSu) [inline server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketOutlierSample.java:79]:
```suggestion
        double value = Math.abs(rowWindow.getRow(i).getInt(0) - avg);
```

File: server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketOutlierSample.java [hunk 6]
```diff
@@ -0,0 +1,1067 @@
...
+        currentValue = rowWindow.getRow(i).getDouble(0);
+        nextValue = rowWindow.getRow(i + 1).getDouble(0);
+
+        x1 = currentTime - lastTime;
+        x2 = nextTime - currentTime;
+        y1 = currentValue - lastValue;
+        y2 = nextValue - currentValue;
+
+        value =
+            (x1 > 0 ? x1 : -x1) + (y1 > 0 ? y1 : -y1) + (x2 > 0 ? x2 : -x2) + (y2 > 0 ? y2 : -y2);
```
Comment (SteveYurongSu) [inline server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketOutlierSample.java:978]: Use Math.abs instead

File: server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketOutlierSample.java [hunk 7]
```diff
@@ -0,0 +1,1067 @@
...
+
+        if (pq.size() < number) {
+          pq.add(new Pair<>(i, value));
+        } else if (value > pq.peek().right) {
+          pq.poll();
+          pq.add(new Pair<>(i, value));
+        }
```
Comment (SteveYurongSu) [inline server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketOutlierSample.java:985]: 我们是不是可以进一步把这段逻辑抽出来？

File: server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketOutlierSample.java [hunk 8]
```diff
@@ -0,0 +1,1067 @@
...
+      int[] arr = new int[number];
+      for (int i = 0; i < number; i++) {
+        arr[i] = pq.peek().left;
+        pq.poll();
+      }
+      Arrays.sort(arr);
+      for (int i = 0; i < number; i++) {
+        collector.putDouble(
+            rowWindow.getRow(arr[i]).getTime(), rowWindow.getRow(arr[i]).getDouble(0));
+      }
```
Comment (SteveYurongSu) [inline server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketOutlierSample.java:997]: 我们是不是可以进一步把这段逻辑抽出来？

File: server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketOutlierSample.java [hunk 9]
```diff
@@ -0,0 +1,1067 @@
...
+
+        lastValue = rowWindow.getRow(i - 1).getDouble(0);
+        currentValue = rowWindow.getRow(i).getDouble(0);
+        nextValue = rowWindow.getRow(i + 1).getDouble(0);
+
+        x1 = currentTime - lastTime;
+        x2 = nextTime - currentTime;
+        y1 = currentValue - lastValue;
+        y2 = nextValue - currentValue;
+
```
Comment (SteveYurongSu) [inline server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketOutlierSample.java:976]:
```suggestion
        x1 = Math.abs(currentTime - lastTime);
        x2 = Math.abs(nextTime - currentTime);
        y1 = Math.abs(currentValue - lastValue);
        y2 = Math.abs(nextValue - currentValue);
        
```

File: server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketOutlierSample.java [hunk 10]
```diff
@@ -0,0 +1,1067 @@
...
+          collector.putInt(row.getTime(), row.getInt(0));
+        }
+        return;
+      } else if (windowSize == number + 2) {
+        for (int i = 1; i < windowSize - 1; i++) {
+          Row row = rowWindow.getRow(i);
+          collector.putInt(row.getTime(), row.getInt(0));
+        }
+        return;
+      }
```
Comment (SteveYurongSu) [inline server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketOutlierSample.java:241]: 这一段是不是可以抽出来

**Review by SteveYurongSu — CHANGES_REQUESTED** [review]

File: docs/UserGuide/Query-Data/Select-Expression.md (hunk 3 above)
Comment (ZhanGHanG9991) [reply docs/UserGuide/Query-Data/Select-Expression.md]: I have modified the docs in zh and eng.

File: docs/zh/UserGuide/Query-Data/Select-Expression.md (hunk 4 above)
Comment (ZhanGHanG9991) [reply docs/zh/UserGuide/Query-Data/Select-Expression.md]: I have modified the docs in zh and eng.

File: server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketOutlierSample.java (hunk 5 above)
Comment (ZhanGHanG9991) [reply server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketOutlierSample.java]: I have fixed this bug.

File: server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketOutlierSample.java (hunk 8 above)
Comment (ZhanGHanG9991) [reply server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketOutlierSample.java]: 已抽出

File: server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketOutlierSample.java (hunk 10 above)
Comment (ZhanGHanG9991) [reply server/src/main/java/org/apache/iotdb/db/query/udf/builtin/UDTFEqualSizeBucketOutlierSample.java]: 已抽出

**Review by SteveYurongSu — APPROVED** [review]

**Pull request merged**